README.md
## leaflet.multiopacity <a href='http://meantrix.com'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->
[![version](https://img.shields.io/badge/version-0.1.1-blue.svg)](https://semver.org)
[![CRAN status](https://www.r-pkg.org/badges/version/leaflet.multiopacity)](https://CRAN.R-project.org/package=leaflet.multiopacity)
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg)](https://opensource.org/licenses/MIT)
[![Codecov test coverage](https://codecov.io/gh/meantrix/leaflet.multiopacity/branch/main/graph/badge.svg)](https://codecov.io/gh/meantrix/leaflet.multiopacity?branch=main)
[![R build status](https://github.com/meantrix/leaflet.multiopacity/workflows/R-CMD-check/badge.svg)](https://github.com/meantrix/leaflet.multiopacity/actions)
<!-- badges: end -->

### Overview

Extends Leaflet for R by adding widget to control opacity of multiple layers, based on **Leaflet.Control.Opacity** ([dayjournal/Leaflet.Control.Opacity](https://github.com/dayjournal/Leaflet.Control.Opacity)) JavaScript plugin with some modifications to support layers other than tile layers (currently tested: tile, image and marker).

### Installation

Before you begin, ensure you have met the following requirement(s):

- You have `R >= 3.5.0` installed.

Install the latest version released on CRAN:

```r
install.packages("leaflet.multiopacity")
```

Or install the development version from GitHub:

```r
if (!require("remotes")) install.packages("remotes")
remotes::install_github("meantrix/leaflet.multiopacity@main")
```

### Basic Usage

`leaflet.multiopacity` provides the function `addOpacityControls` to include opacity sliders in leaflet map.

```r
# Load libraries
library(leaflet)
library(leaflet.multiopacity)
library(raster)

# Create raster example
r <- raster(xmn = -2.8, xmx = -2.79,
            ymn = 54.04, ymx = 54.05,
            nrows = 30, ncols = 30)
values(r) <- matrix(1:900, nrow(r), ncol(r), byrow = TRUE)
crs(r) <- crs("+init=epsg:4326")

# Provide layerId, group or category to show opacity controls
# If not specified, will render controls for all layers
leaflet() %>%
  addProviderTiles("Wikimedia", layerId = "Wikimedia") %>%
  addRasterImage(r, layerId = "raster") %>%
  addAwesomeMarkers(lng = -2.79545, lat = 54.04321,
                    layerId = "hospital", label = "Hospital") %>%
  addOpacityControls(layerId = c("raster", "hospital"),
                     collapsed = FALSE, position = "topright",
                     title = "Opacity Control")
```

For more information and examples, please take a look at the vignettes:

[Usage with leaflet](https://meantrix.github.io/leaflet.multiopacity/articles/usage-leaflet.html)  

[Usage with leafletProxy](https://meantrix.github.io/leaflet.multiopacity/articles/usage-leafletProxy.html)

### Contributing to leaflet.multiopacity

To contribute to `leaflet.multiopacity`, follow these steps:

1. Fork this repository.
2. Create a branch: `git checkout -b <branch_name>`.
3. Make your changes and commit them: `git commit -m '<commit_message>'`
4. Push to the original branch: `git push origin leaflet.multiopacity/<location>`
5. Create the pull request.

Alternatively see the GitHub documentation on [creating a pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request).

### Bug Reports

If you have detected a bug (or want to ask for a new feature), please file an issue with a minimal [reproducible example](https://www.tidyverse.org/help/#reprex) on [GitHub](https://github.com/meantrix/leaflet.multiopacity/issues).

### License

This project uses the following license: [MIT License](<https://github.com/meantrix/leaflet.multiopacity/blob/master/LICENSE>).


NEWS.md
# leaflet.multiopacity 0.1.1

* Fix a bug with control icon when collapsed is TRUE.
* Implement a new argument to set control size when collapsed is enabled.

# leaflet.multiopacity 0.1.0

* Initial release of `leaflet.multiopacity`.


tests/testthat.R
library(testthat)
library(leaflet.multiopacity)

test_check("leaflet.multiopacity")


tests/testthat/test-addOpacityControls.R
test_that("addOpacityControls dependencies are included in leaflet map", {
  map <- addOpacityControls(leaflet::leaflet())
  expect_equal(map$dependencies[[1]]$name, "Leaflet.Control.Opacity")
  expect_equal(map$dependencies[[2]]$name, "leafletR.utils")
})

test_that("addOpacityControls JS function is registered on jsHooks", {
  map <- addOpacityControls(leaflet::leaflet())
  map2 <- addOpacityControls(leaflet::leaflet(), renderOnLayerAdd = TRUE)
  expect_match(map$jsHooks$render[[1]]$code[[1]], "function\\(el, x, data\\).*")
  expect_match(map2$jsHooks$render[[1]]$code[[1]], "function\\(el, x, data\\).*")
})


R/utils.R
#' Return JavaScript dependencies
#'
#' @return List with JavaScript dependencies of the package.
dependencies <- function() {
  Leaflet.Control.Opacity <- htmltools::htmlDependency(
    name = "Leaflet.Control.Opacity",
    version = "1.4",
    src = system.file("src/Leaflet.Control.Opacity/",
                      package = "leaflet.multiopacity"),
    script = "L.Control.Opacity.js",
    stylesheet = "L.Control.Opacity.css"
  )
  leafletR.utils <- htmltools::htmlDependency(
    name = "leafletR.utils",
    version = "0.1.0",
    src = system.file("src/leafletR.utils/",
                      package = "leaflet.multiopacity"),
    script = "leafletR.utils.js"
  )
  return(list(Leaflet.Control.Opacity, leafletR.utils))
}

#' Register Plugin to leaflet map
#'
#' @param map
#' The map to register the plugin.
#' @param plugin
#' List with JavaScript dependencies (from 'htmltools::htmlDependency')
#' to register.
#'
#' @return
#' Leaflet map with plugin registered (JS dependencies added).
registerPlugin <- function(map, plugin) {
  map$dependencies <- c(map$dependencies, plugin)
  return(map)
}

#' Leaflet Debug
#'
#' Debug a leaflet map in client side using browser inspector tools.
#'
#' @param map Leaflet map.
#'
#' @return Leaflet map.
leafletDebug <- function(map) {
  htmlwidgets::onRender(
    map,
    jsCode = htmlwidgets::JS("function(el, x) {
                              var map = this;
                              debugger;
                             }")
  )
}


R/dependencies.R
#' Return JavaScript dependencies
#'
#' @return List with JavaScript dependencies of the package.
dependencies <- function() {
  Leaflet.Control.Opacity <- htmltools::htmlDependency(
    name = "Leaflet.Control.Opacity",
    version = "1.4",
    src = system.file("src/Leaflet.Control.Opacity/",
                      package = "leaflet.multiopacity"),
    script = "L.Control.Opacity.js",
    stylesheet = "L.Control.Opacity.css"
  )
  leafletR.utils <- htmltools::htmlDependency(
    name = "leafletR.utils",
    version = "0.1.0",
    src = system.file("src/leafletR.utils/",
                      package = "leaflet.multiopacity"),
    script = "leafletR.utils.js"
  )
  return(list(Leaflet.Control.Opacity, leafletR.utils))
}

#' Register Plugin to leaflet map
#'
#' @param map
#' The map to register the plugin.
#' @param plugin
#' List with JavaScript dependencies (from 'htmltools::htmlDependency')
#' to register.
#'
#' @return
#' Leaflet map with plugin registered (JS dependencies added).
registerPlugin <- function(map, plugin) {
  map$dependencies <- c(map$dependencies, plugin)
  return(map)
}


inst/src/leafletR.utils/leafletR.utils.js
// #########################
// Utility Functions
// Meantrix
// #########################

// Coerce value into array
function asArray(value) {
  if (value === null)
    return [];
  if (Array.isArray(value))
    return value;
  return [value];
}

// Get all layers from map.layerManager._byStamp
function getAllLayers(ObjectByStamp) {
  var layerObjs = {};
  for (var i = 0; i < Object.keys(ObjectByStamp).length; i++) {
    var layer = Object.values(ObjectByStamp)[i];
    var layerId;
    if (layer.layerId === null) {
      layerId = layer.layer._leaflet_id;
    } else {
      layerId = layer.layerId;
    }
    var layerObj = {
      layerId: layerId, //layer.layerId,
      category: layer.category,
      group: layer.group,
      _leaflet_id: layer.layer._leaflet_id,
      layer: layer.layer
    };
    layerObjs[layerId] = layerObj;
  }
  return(layerObjs);
}

// Detect if text has some of searchWords
function multiSearchOr(text, searchWords) {
  if (text == null) return false;
  match = searchWords.map(el => "\\b" + el + "\\b").some(function(el) {
    return(text.match(new RegExp(el,"i")));
  });
  return match;
}

// Detect if text has all the searchWords
function multiSearchAnd(text, searchWords) {
  if (text == null) return false;
  match = searchWords.map(el => "\\b" + el + "\\b").every(function(el) {
    return(text.match(new RegExp(el,"i")));
  });
  return match;
}

// Subset allLayers by layerId(s)
function subsetByLayerId(allLayers, layerId) {
  var res;
  if (layerId === null) {
    res = Object.values(allLayers);
  } else {
    // Filter by layerId
    res = Object.values(allLayers).filter(
      function(p) {return(multiSearchOr(p.layerId, asArray(layerId)))}
    );
  }
  obj = {};
  for (i = 0; i < res.length; i++) {
    obj[res[i].layerId] = res[i].layer;
  }
  return(obj);
}

// Subset allLayers by group(s)
function subsetByGroup(allLayers, group) {
  var res;
  if (group === null) {
    res = Object.values(allLayers);
  } else {
    // Filter by group
    res = Object.values(allLayers).filter(
      function(p) {return(multiSearchOr(p.group, asArray(group)))}
    );
  }
  obj = {};
  for (i = 0; i < res.length; i++) {
    obj[res[i].layerId] = res[i].layer;
  }
  return(obj);
}

// Subset allLayers by category(ies)
function subsetByCategory(allLayers, category) {
  var res;
  if (category === null) {
    res = Object.values(allLayers);
  } else {
    // Filter by category
    res = Object.values(allLayers).filter(
      function(p) {return(multiSearchOr(p.category, asArray(category)))}
    );
  }
  obj = {};
  for (i = 0; i < res.length; i++) {
    obj[res[i].layerId] = res[i].layer;
  }
  return(obj);
}



inst/src/Leaflet.Control.Opacity/L.Control.Opacity.js
L.Control.Opacity = L.Control.extend({
	options: {
		collapsed: false,
		position: 'topright',
		size: 'm',
		label: null
	},
	initialize: function (overlays, options) {
		L.Util.setOptions(this, options);
		this._layerControlInputs = [];
		this._layers = [];
		this._lastZIndex = 0;
		this._handlingClick = false;
		for (var i in overlays) {
			this._addLayer(overlays[i], i, true);
		}
	},
	onAdd: function (map) {
		this._initLayout();
		this._update();
		return this._container;
	},
	expand: function () {
		L.DomUtil.addClass(this._container, 'leaflet-control-layers-expanded');
		this._form.style.height = null;
		var acceptableHeight = this._map.getSize().y - (this._container.offsetTop + 50);
		if (acceptableHeight < this._form.clientHeight) {
			L.DomUtil.addClass(this._form, 'leaflet-control-layers-scrollbar');
			this._form.style.height = acceptableHeight + 'px';
		} else {
			L.DomUtil.removeClass(this._form, 'leaflet-control-layers-scrollbar');
		}
		return this;
	},
	collapse: function () {
		L.DomUtil.removeClass(this._container, 'leaflet-control-layers-expanded');
		return this;
	},
	_initLayout: function () {
		var className = 'leaflet-control-layers',
		    container = this._container = L.DomUtil.create('div', className),
		    collapsed = this.options.collapsed;
		    size = this.options.size;
		container.setAttribute('aria-haspopup', true);
		L.DomEvent.disableClickPropagation(container);
		L.DomEvent.disableScrollPropagation(container);
		if(this.options.label){
			var labelP = L.DomUtil.create('p', className + "-label");
			labelP.innerHTML = this.options.label;
			container.appendChild(labelP);
		}
		var form = this._form = L.DomUtil.create('form', className + '-list');
		if (collapsed) {
			this._map.on('click', this.collapse, this);
			if (!L.Browser.android) {
				L.DomEvent.on(container, {
					mouseenter: this.expand,
					mouseleave: this.collapse
				}, this);
			}
		}
		//var link = this._layersLink = L.DomUtil.create('a', className + '-toggle', container);
		var link = this._layersLink = L.DomUtil.create('a', className + '-toggle multiopacity-' + size, container);
		link.href = '#';
		link.title = 'Layers';
		if (L.Browser.touch) {
			L.DomEvent.on(link, 'click', L.DomEvent.stop);
			L.DomEvent.on(link, 'click', this.expand, this);
		} else {
			L.DomEvent.on(link, 'focus', this.expand, this);
		}
		if (!collapsed) {
			this.expand();
		}
		this._baseLayersList = L.DomUtil.create('div', className + '-base', form);
		this._separator = L.DomUtil.create('div', className + '-separator', form);
		this._overlaysList = L.DomUtil.create('div', className + '-overlays', form);
		container.appendChild(form);
	},
	_getLayer: function (id) {
		for (var i = 0; i < this._layers.length; i++) {
			if (this._layers[i] && L.Util.stamp(this._layers[i].layer) === id) {
				return this._layers[i];
			}
		}
	},
	_addLayer: function (layer, name, overlay) {
		this._layers.push({
			layer: layer,
			name: name,
			overlay: overlay
		});
	},
	_update: function () {
		if (!this._container) { return this; }
		L.DomUtil.empty(this._baseLayersList);
		L.DomUtil.empty(this._overlaysList);
		this._layerControlInputs = [];
		var baseLayersPresent, overlaysPresent, i, obj, baseLayersCount = 0;
		for (i = 0; i < this._layers.length; i++) {
			obj = this._layers[i];
			this._addItem(obj);
			overlaysPresent = overlaysPresent || obj.overlay;
			baseLayersPresent = baseLayersPresent || !obj.overlay;
			baseLayersCount += !obj.overlay ? 1 : 0;
		}
		if (this.options.hideSingleBase) {
			baseLayersPresent = baseLayersPresent && baseLayersCount > 1;
			this._baseLayersList.style.display = baseLayersPresent ? '' : 'none';
		}
		this._separator.style.display = overlaysPresent && baseLayersPresent ? '' : 'none';
		return this;
	},
	_addItem: function (obj) {
		var label = document.createElement('label'),
		    input;
		if (obj.overlay) {
			input = document.createElement('input');
			input.type = 'range';
			input.className = 'leaflet-control-layers-range';
            input.min = 0;
            input.max = 100;
            //input.value = 100;
            input.value = obj.layer.options.opacity * 100;
		} else {
			input = this._createRadioElement('leaflet-base-layers', checked);
		}
		this._layerControlInputs.push(input);
		input.layerId = L.Util.stamp(obj.layer);
		L.DomEvent.on(input, 'click', this._onInputClick, this);
		L.DomEvent.on(input, 'touchend', this._onInputClick, this);
		var name = document.createElement('span');
		name.innerHTML = ' ' + obj.name;
		var holder = document.createElement('div');
		var holder2 = document.createElement('div');
		label.appendChild(holder);
		holder.appendChild(name);
		label.appendChild(holder2);
		holder2.appendChild(input);
		var container = obj.overlay ? this._overlaysList : this._baseLayersList;
		container.appendChild(label);
		return label;
	},
	_onInputClick: function () {
		var inputs = this._layerControlInputs,
			input, layer;

		this._handlingClick = true;

		for (var i = inputs.length - 1; i >= 0; i--) {
			input = inputs[i];
			layer = this._getLayer(input.layerId).layer;
			//if( typeof layer._url === 'undefined'){ // disable url check
			//} else {
			//layer.options.fillOpacity = input.value/100
			//layer.options.opacity = input.value/100
			layer.setOpacity(input.value / 100);
			//}
		}

		this._handlingClick = false;

		this._refocusOnMap();
	},
});

L.control.opacity = function (overlays, options) {
        return new L.Control.Opacity(overlays, options);
};


R/multiopacity.R
#' Add Opacity Controls
#'
#' Add opacity controls to leaflet map. It is possible to choose which layers will
#' have an opacity slider control by passing one of the arguments "category", "group"
#' or "layerId" (passing more than one of these arguments will cause the others to
#' be ignored). It is possible to pass one or more of each layerIds, categories or
#' groups. If you leave these arguments with the default value (NULL), the opacity
#' controls will be created to every layer in the map.
#'
#' @param map
#' The map to add the opacity controls to.
#' @param category
#' One or more categories to render opacity controls to.
#' Tested categories are "tile" (base layers), "image"
#' (e.g. raster) and "marker".
#' @param group
#' One or more groups to render opacity controls to.
#' @param layerId
#' One or more layer IDs to render opacity controls to.
#' @param collapsed
#' If FALSE (the default), the opacity control will always appear
#' in its expanded state. Set to TRUE to have the opacity control
#' rendered as an icon that expands when hovered over.
#' @param size
#' Collapsed control size: "m" (medium) or "s" (small).
#' @param position
#' Position of control: "topleft", "topright", "bottomleft", or "bottomright".
#' @param title
#' The control title.
#' @param renderOnLayerAdd
#' When this argument is TRUE, the controls will only appear when a new
#' layer is added and rendered in the map. This can be useful if you plan
#' to use 'leafletProxy()' and need the controls to be dinamically updated.
#'
#' @examples
#' # Load libraries
#' library(leaflet)
#' library(leaflet.multiopacity)
#' library(raster)
#'
#' # Create raster example
#' r <- raster(xmn = -2.8, xmx = -2.79,
#'             ymn = 54.04, ymx = 54.05,
#'             nrows = 30, ncols = 30)
#' values(r) <- matrix(1:900, nrow(r), ncol(r), byrow = TRUE)
#' crs(r) <- crs("+init=epsg:4326")
#'
#' # Provide layerId, group or category to show opacity controls
#' # If not specified, will render controls for all layers
#' # Example using layerId
#' leaflet() %>%
#'   addProviderTiles("Wikimedia", layerId = "Wikimedia") %>%
#'   addRasterImage(r, layerId = "raster") %>%
#'   addAwesomeMarkers(lng = -2.79545, lat = 54.04321,
#'                     layerId = "hospital", label = "Hospital") %>%
#'   addOpacityControls(layerId = c("raster", "hospital"))
#'
#' # Example using category
#' leaflet() %>%
#'   addProviderTiles("Wikimedia", layerId = "Wikimedia") %>%
#'   addRasterImage(r, layerId = "raster") %>%
#'   addAwesomeMarkers(lng = -2.79545, lat = 54.04321,
#'                     layerId = "hospital", label = "Hospital") %>%
#'   addOpacityControls(category = c("image", "marker"))
#'
#' @export
addOpacityControls <- function(map,
                               layerId = NULL,
                               category = NULL,
                               group = NULL,
                               collapsed = FALSE,
                               size = c("m", "s"),
                               position = c(
                                 "topright",
                                 "topleft",
                                 "bottomright",
                                 "bottomleft"
                               ),
                               title = NULL,
                               renderOnLayerAdd = FALSE) {

  if (inherits(map, "leaflet_proxy"))
    stop("LeafletProxy is not supported. Please use this function with renderOnLayerAdd set to TRUE.")
  stopifnot(inherits(map, "leaflet"))
  if (!is.null(layerId)) stopifnot(inherits(layerId, "character"))
  if (!is.null(category)) stopifnot(inherits(category, "character"))
  if (!is.null(group)) stopifnot(inherits(group, "character"))
  stopifnot(inherits(collapsed, "logical"), length(collapsed) == 1)
  if (!is.null(title)) stopifnot(inherits(title, "character"), length(title) == 1)
  stopifnot(inherits(renderOnLayerAdd, "logical"), length(renderOnLayerAdd) == 1)
  size <- match.arg(size)
  position <- match.arg(position)

  if (isTRUE(collapsed) && !is.null(title)) {
    warning("Not possible to set control title when collapsed is TRUE.")
    title <- NULL
  }

  type <- .chooseType(layerId = layerId, category = category, group = group)

  data <- list(category = category,
               group = group,
               layerId = layerId,
               options = list(
                 type = type,
                 collapsed = collapsed,
                 position = position,
                 size = size,
                 label = title
               ))

  map <- registerPlugin(map, dependencies())

  if (renderOnLayerAdd) {
    jsCode <- htmlwidgets::JS('
      function(el, x, data) {
        var multiopacityControl;
        var map = this;
        map.on("layeradd",
          function(e) {
            // remove previous multiopacityControl if it exists
            if (multiopacityControl !== undefined) {
              multiopacityControl.remove()
            }
            var allLayers = getAllLayers(map.layerManager._byStamp)
            var layers;
            switch (data.options.type) {
              case "category":
                layers = subsetByCategory(allLayers, data.category);
                break;
              case "group":
                layers = subsetByGroup(allLayers, data.group);
                break;
              case "layerId":
                layers = subsetByLayerId(allLayers, data.layerId);
                break;
              default:
                layers = subsetByLayerId(allLayers, null);
                break;
            }
            //OpacityControl
              multiopacityControl = L.control.opacity(
                layers,
                options = {
                  collapsed: data.options.collapsed,
                  position: data.options.position,
                  size: data.options.size,
                  label: data.options.label
                }
              ).addTo(map);
          }
        )
      }')
  } else {
    jsCode <- htmlwidgets::JS('
      function(el, x, data) {
        var multiopacityControl;
        var map = this;
        // remove previous multiopacityControl if it exists
        if (multiopacityControl !== undefined) {
          multiopacityControl.remove()
        }
        var allLayers = getAllLayers(map.layerManager._byStamp)
        switch (data.options.type) {
          case "category":
            var layers = subsetByCategory(allLayers, data.category);
            break;
          case "group":
            var layers = subsetByGroup(allLayers, data.group);
            break;
          case "layerId":
            var layers = subsetByLayerId(allLayers, data.layerId);
            break;
          default:
            var layers = subsetByLayerId(allLayers, null);
            break;
        }
        //OpacityControl
        multiopacityControl = L.control.opacity(
          layers,
          options = {
            collapsed: data.options.collapsed,
            position: data.options.position,
            size: data.options.size,
            label: data.options.label
          }
        ).addTo(map);
      }')
  }

  htmlwidgets::onRender(
    map,
    jsCode = jsCode,
    data = data)

}

.chooseType <- function(layerId = NULL, category = NULL, group = NULL) {
  idx <- sapply(list(layerId = layerId,
                     category = category,
                     group = group),
                function(x) !is.null(x))
  type <- names(which(idx))
  if (length(type) == 0) {
    message("No filter used, showing all layers.")
    type <- "none"
  } else if (length(type) > 1) {
    type <- type[1]
    warning(
      paste("It is not possible to pass more than",
            "one type (layerId, category or group).",
            "Using", type, "only."),
      call. = FALSE)
  }
  return(type)
}
